### PR TITLE
refactor: add more error info when array is empty

### LINKED ISSUE #5557: median on empty input returns confusing error
**Describe the bug**

If trying to get median on empty input/dataframe a confusing error is returned

**To Reproduce**

```sql
DataFusion CLI v20.0.0
❯ select median(a) from (select 1 as a where 1=0);
Internal error: Empty iterator passed to ScalarValue::iter_to_array. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
❯
```

**Expected behavior**

Either return some sensible value (e.g. null) or a more informative error message

**Additional context**

This also affects the dataframe `describe(...)` method which uses median

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #5557

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/median.rs [hunk 1]
```diff
@@ -151,7 +151,16 @@ impl Accumulator for MedianAccumulator {
                 // ignore null values
                 .filter(|v| !v.is_null())
                 .cloned(),
-        )?;
+        );
+
+        if array.is_err() {
```
Comment (comphead) [inline datafusion/physical-expr/src/aggregate/median.rs:156]: `array.map_err(|e| DataFusionError::Execution("".to_owned()))?;` You may find this syntax more concise

File: datafusion/physical-expr/src/aggregate/approx_median.rs [hunk 2]
```diff
@@ -67,7 +67,11 @@ impl AggregateExpr for ApproxMedian {
     }
 
     fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
-        self.approx_percentile.create_accumulator()
+        self.approx_percentile.create_accumulator().map_err(|_| {
+            DataFusionError::Execution(
+                "approx median requires at least one non-null value".to_string(),
+            )
+        })
```
Comment (Jefffrey) [inline datafusion/physical-expr/src/aggregate/approx_median.rs:74]: I'm not sure if this wrapping will prevent having an empty input? Could you add some tests to check this?

File: datafusion/physical-expr/src/aggregate/median.rs [hunk 3]
```diff
@@ -151,7 +151,12 @@ impl Accumulator for MedianAccumulator {
                 // ignore null values
                 .filter(|v| !v.is_null())
                 .cloned(),
-        )?;
+        )
+        .map_err(|_| {
+            DataFusionError::Execution(
+                "median requires at least one non-null value".into(),
+            )
+        })?;
```
Comment (Jefffrey) [inline datafusion/physical-expr/src/aggregate/median.rs:159]: Does this have the possibility of incorrectly overwriting other errors that might have occurred in `iter_to_array`?

**Review by Jefffrey — COMMENTED** [review]
Thanks for picking this up, left some comments, and some tests would be nice too 🙂

Though worth noting that polars seems to return null for empty inputs:

```python
>>> df = pl.DataFrame({"foo": []})
>>> df.median()
shape: (1, 1)
┌──────┐
│ foo  │
│ ---  │
│ f32  │
╞══════╡
│ null │
└──────┘
>>>
```

Can see what other people might think is the preferred behaviour (error vs null).

File: datafusion/physical-expr/src/aggregate/median.rs (hunk 3 above)
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/median.rs:159]: I think so.

What about explicitly checking for all nulls with something like:

```rust
if  !self.all_values
                .iter()
                .any(|v| !v.is_null()) {
.. raise error here
}
```

**Review by alamb — COMMENTED** [review]
Thank you for the contribution @Weijun-H  and the review @Jefffrey 

@Jefffrey  makes some excellent points and I think this PR should have some tests showing the new errors -- perhaps added to https://github.com/apache/arrow-datafusion/tree/main/datafusion/core/tests/sqllogictests

**Review by alamb — APPROVED** [review]
Thank you @Weijun-H  -- looks great to me

**Comment by alamb** [comment]
> Though worth noting that polars seems to return null for empty inputs:

I agree null would be better -- filed https://github.com/apache/arrow-datafusion/issues/5595

But this PR is better than what is on main, so merging it in

**Pull request merged**